src/NumericalAlgorithms/LinearOperators/ApplyMatricesToVariables.cpp

```diff
@@ -0,0 +1,365 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/LinearOperators/ApplyMatricesToVariables.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <vector>
+
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "NumericalAlgorithms/LinearOperators/Transpose.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace {
+void multiply_in_first_dimension(const Matrix& matrix, const double* data,
+                                 const gsl::not_null<size_t*> data_size,
+                                 const gsl::not_null<double*> result) noexcept {
+  *data_size /= matrix.columns();
+  dgemm_<true>('N', 'N',
+               matrix.rows(),     // rows matrix and result
+               *data_size,        // columns result and u
+               matrix.columns(),  // columns matrix and rows u
+               1.0,               // overall multiplier
+               matrix.data(),     // matrix
+               matrix.rows(),     // ...and rows
+               data,              // u
+               matrix.columns(),  // ..and rows
+               0.0,               // multiplier for unused term
+               result.get(),      // result
+               matrix.rows());    // ...and rows
+  *data_size *= matrix.rows();
+}
+
+void do_transpose(const double* const data, const size_t data_size,
+                  const size_t chunk_size,
+                  const gsl::not_null<double*> result) noexcept {
+  raw_transpose(data, chunk_size, data_size / chunk_size, result);
+}
+
+struct Scratch {
+  std::vector<double> buffer;
+  double* a;
+  double* b;
+};
+
+template <typename MatrixType, size_t Dim>
```

Comment:
add comment about that this is a rough, but guaranteed to be sufficient, estimate of scratch space needed